k8s/controllers/solution/instance_queueing_controller.go

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ * SPDX-License-Identifier: MIT
+ */
+
+package solution
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	fabric_v1 "gopls-workspace/apis/fabric/v1"
+	solution_v1 "gopls-workspace/apis/solution/v1"
+	"gopls-workspace/constants"
+	"gopls-workspace/controllers/metrics"
+	"gopls-workspace/predicates"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// InstanceQueueingReconciler reconciles a Instance object
+type InstanceQueueingReconciler struct {
+	InstanceReconciler
+}
+
+//+kubebuilder:rbac:groups=solution.symphony,resources=instances,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=solution.symphony,resources=instances/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=solution.symphony,resources=instances/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Instance object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *InstanceQueueingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Reconcile Queueing Instance " + req.Name + " in namespace " + req.Namespace)
+
+	// Initialize reconcileTime for latency metrics
+	reconcileTime := time.Now()
+
+	// Get instance
+	instance := &solution_v1.Instance{}
+	if err := r.Client.Get(ctx, req.NamespacedName, instance); err != nil {
+		log.Error(err, "unable to fetch Instance object")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	reconciliationType := metrics.CreateOperationType
+	resultType := metrics.ReconcileSuccessResult
+	reconcileResult := ctrl.Result{}
+	deploymentOperationType := metrics.DeploymentQueued
+	var err error
+
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() { // update
+		reconciliationType = metrics.UpdateOperationType
+		operationName := fmt.Sprintf("%s/%s", constants.InstanceOperationNamePrefix, constants.ActivityOperation_Write)
+		deploymentOperationType, reconcileResult, err = r.dr.AttemptUpdate(ctx, instance, false, log, instanceOperationStartTimeKey, operationName)
```

Comment:
If AttemptUpdate fails the job will be re-queued by K8s, right? Where's the de-dup logic for it? Consider the case where a target is turned off or has some unrecoverable errors, this is the case where we will accumulate many jobs that can't be finished correctly.

Reply:
In this case, the queueing job part should be successful. So the error only happended during the job processing in the symphony api. And if the job is failling at symphony api side. Symphony api has the logic to redo the job. but it shoun't impact the queueing, since api is good